File: native-plugins/admob/resources/android/src/RewardedRegistry.kt
package com.blessedzulu.nativephp.admob

import com.google.android.gms.ads.rewarded.RewardedAd

/**
 * Slot-keyed registry for loaded RewardedAd instances.
 *
 * Same one-shot semantics as InterstitialRegistry: each loaded ad survives
 * until shown (and dismissed, with or without earning a reward) or failed
 * to show. The slot is cleared in those terminal callbacks.
 *
 * Thread-safety: all public methods are synchronized. Callers should still
 * post() to the main thread before invoking RewardedAd APIs since Google's
 * SDK requires it.
 */
object RewardedRegistry {
    private val ads = mutableMapOf<String, RewardedAd>()

    @Synchronized
    fun put(slot: String, ad: RewardedAd) {
        ads[slot] = ad
    }

    @Synchronized
    fun get(slot: String): RewardedAd? = ads[slot]

    @Synchronized
    fun remove(slot: String): RewardedAd? = ads.remove(slot)

    @Synchronized
    fun clear() {
        ads.clear()
    }
}

File: native-plugins/admob/resources/ios/Sources/RewardedRegistry.swift
import Foundation
import GoogleMobileAds

/// Slot-keyed registry for loaded `RewardedAd` instances and their delegates.
///
/// Same as `InterstitialRegistry`: `FullScreenContentDelegate` is weakly held
/// by `RewardedAd`, so the registry strongly retains the delegate alongside the
/// ad until dismissal. Without that the delegate is deallocated before
/// `present(from:userDidEarnRewardHandler:)` fires its callbacks.
///
/// One-shot: the slot is cleared in dismissal or failed-show callbacks.
final class RewardedRegistry {
    static let shared = RewardedRegistry()

    private var ads: [String: RewardedAd] = [:]
    private var delegates: [String: RewardedDelegate] = [:]

    private init() {}

    func put(slot: String, ad: RewardedAd, delegate: RewardedDelegate) {
        ads[slot] = ad
        delegates[slot] = delegate
    }

    func get(slot: String) -> RewardedAd? {
        return ads[slot]
    }

    func remove(slot: String) {
        ads.removeValue(forKey: slot)
        delegates.removeValue(forKey: slot)
    }

    func clear() {
        ads.removeAll()
        delegates.removeAll()
    }
}
